### ISSUE TITLE
Tried to map bad logical plan. LHS and RHS must never be the same

### ISSUE THREAD

**Issue by sinanoezdemir** [neo4j/neo4j#12195]
## Tried to map bad logical plan

I am receiving a database error by executing the following cypher statement. 
 
**Neo4j Version:** 3.5.0  
**Operating System:** Docker -> Alpine Linux 3.9.2  
**API:** Cypher

### Steps to reproduce
1. Fill the database with a domain specific data. As of right know I can not provide that data here. 
2. Execute query: 
MATCH (a:SomeLabel)<-[relA:someRelation]-(c)
MATCH (b:SomeLabel)<-[relB:someRelation]-(c)
  WHERE (a.someStringId = b.someStringId AND relA.someIntValue < relB.someIntValue)
  OR a.someStringId < b.someStringId 
return *
3. Logs: [modofied_log.txt](https://github.com/neo4j/neo4j/files/3053856/modofied_log.txt)

### Additional Information

- Wrapping the values in the OR statement with toString() resolves that issue. 
MATCH (a:SomeLabel)<-[relA:someRelation]-(c)
MATCH (b:SomeLabel)<-[relB:someRelation]-(c)
  WHERE (a.someStringId = b.someStringId AND relA.someIntValue < relB.someIntValue)
  OR **_toString(a.someStringId) < toString(b.someStringId)_** 
return * 
- This error only occurred by having more than 2000 nodes and 4000 relationships.

**Comment by sherfert** [comment]
Thanks for the bug report. We'll look into it.

**Comment by sherfert** [comment]
@sinanoezdemir By simply creating the right amount of nodes, relationships, and the index that I see must exist from the log, I cannot reproduce the error. I am getting a different plan.

Could you run

```
curl -H accept:application/json -H content-type:application/json -d '{"statements":[{"statement":"CALL db.stats.retrieve(\"GRAPH COUNTS\")"}]}' http://_________:7474/db/data/transaction/commit > graphCounts.json
```

and share that json file (after anonymizing it)? It will contain statistical information that the planner has access to, so we can use that to reproduce the same code path.

Edit: I see you are on 3.5.0, I think that was only added in 3.5.1. Is it possible for you to update to a later 3.5?

**Comment by zaferad** [comment]
Hi @sherfert,

I've run the procedure against different setups in our database. The bug comes up with 

| Neo4J | APOC | Graph Algorithms |
| --- | --- | --- |
| 3.5.0 | 3.5.0.1 | 3.5.0.1 |
| 3.5.1 | 3.5.0.1 | 3.5.1.0 |
| 3.5.2 | 3.5.0.1 | 3.5.2.0 |
| 3.5.3 | 3.5.0.3 | 3.5.4.0 |
| 3.5.4 | 3.5.0.3 | 3.5.4.0 |

The APOC and Graph Algo versions were suggested by Neo4J Desktop 1.1.17.

The bug seems to disappear with the following setup:

| Neo4J | APOC | Graph Algorithms |
| --- | --- | --- |
| 3.5.4 | 3.5.0.2 | 3.5.4.0 |

APOC was manually downloaded from Github at https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/tag/3.5.0.2

In all setups, we include the oracle database driver jar and a custom neo4j-plugin jar with the same versions.

[graphCounts.json.txt](https://github.com/neo4j/neo4j/files/3107330/graphCounts.json.txt) (made with 3.5.1)
[graphCounts3.5.4.json.txt](https://github.com/neo4j/neo4j/files/3107331/graphCounts3.5.4.json.txt) (using the "working" setup)

**Comment by sherfert** [comment]
We have merged a fix for the issue which will be released in 3.5.6. Thanks again for reporting it!
